### PR TITLE
Feature/change log tools improvement

### PR DESCRIPTION
**Pull request by chidozieononiwu**
Update ChangeLog Logic

### CODE REVIEW

File: eng/common/scripts/Update-ChangeLog.ps1 [hunk 1]
```diff
@@ -0,0 +1,109 @@
...
+  [String]$Version,
+  [Parameter(Mandatory = $true)]
+  [String]$ServiceDirectory,
+  [Parameter(Mandatory = $true)]
+  [String]$PackageName,
+  [boolean]$Unreleased=$True,
+  [boolean]$ReplaceVersion = $False
+)
+
+DynamicParam {
```
Comment (weshaggard) [inline eng/common/scripts/Update-ChangeLog.ps1:20]: What is a DynamicParam? Why is this needed?

File: eng/common/scripts/Update-ChangeLog.ps1 [hunk 2]
```diff
@@ -0,0 +1,109 @@
+# Note: This script will add or replace version title in change log
```
Comment (weshaggard) [inline eng/common/scripts/Update-ChangeLog.ps1:1]: I assume this script will replace https://github.com/Azure/azure-sdk-tools/blob/master/eng/common/Update-Change-Log.ps1 and you will delete that once you transition folks to this one?

File: eng/common/scripts/Update-ChangeLog.ps1 [hunk 3]
```diff
@@ -0,0 +1,109 @@
...
+        $attributeCollection = New-object System.Collections.ObjectModel.Collection[System.Attribute]
+        $attributeCollection.Add($releaseStatusAttribute)
+        $releaseStatusParam = New-Object System.Management.Automation.RuntimeDefinedParameter('ReleaseDate', [string], $attributeCollection)
+        $paramDictionary = New-Object System.Management.Automation.RuntimeDefinedParameterDictionary
+        $paramDictionary.Add('ReleaseDate', $releaseStatusParam)
+        return $paramDictionary
+    }
+}
+
+Begin {
```
Comment (weshaggard) [inline eng/common/scripts/Update-ChangeLog.ps1:33]: Why the Begin and Process blocks?

File: eng/common/scripts/copy-docs-to-blobstorage.ps1 [hunk 4]
```diff
@@ -231,7 +231,7 @@ function Upload-Blobs
 }
 
 
-if ((Get-ChildItem -Path Function: | ? { $_.Name -eq $PublishGithubIODocsFn }).Count -gt 0)
+if (Test-Path Function:PublishGithubIODocsFn)
```
Comment (weshaggard) [inline eng/common/scripts/copy-docs-to-blobstorage.ps1:234]: I like this improvement it makes these checks a lot cleaner.

File: eng/pipelines/eng-common-sync.yml [hunk 5]
```diff
@@ -36,7 +36,7 @@ pr:
       - eng/common
 
 pool:
-  vmImage: windows-2019
+  vmImage: ubuntu-18.04
```
Comment (weshaggard) [inline eng/pipelines/eng-common-sync.yml:39]: Is there a particular reason for this switch?

File: eng/common/scripts/Update-ChangeLog.ps1 [hunk 6]
```diff
@@ -0,0 +1,109 @@
...
+    . "${PSScriptRoot}\common.ps1"
+    $UNRELEASED_TAG = "(Unreleased)"
```
Comment (weshaggard) [inline eng/common/scripts/Update-ChangeLog.ps1:35]: Why isn't this in changelog operations?

File: eng/common/scripts/ChangeLog-Operations.ps1 [hunk 7]
```diff
@@ -121,10 +119,117 @@ function Confirm-ChangeLogEntry {
...
-# Release History
-## $TestVersion ($(Get-Date -f "yyyy-MM-dd"))
-- $ReleaseEntry
-"@
+function New-ChangeLogEntry {
+  param (
+    [Parameter(Mandatory = $true)]
+    [ValidateNotNullOrEmpty()]
+    [String]$Version,
+    [String]$Status="(Unreleased)",
```
Comment (weshaggard) [inline eng/common/scripts/ChangeLog-Operations.ps1:127]: Looks like we should be using the constant for `Unreleased` here and in other places.

File: eng/common/scripts/ChangeLog-Operations.ps1 [hunk 8]
```diff
@@ -121,10 +119,117 @@ function Confirm-ChangeLogEntry {
...
+    [String[]]$Content
+  )
+
+  # Validate relase Status
+  $Status = $Status.Trim().Trim("()")
+  if ($Status -ne "Unreleased") {
+    $dateFormat = "yyyy-MM-dd"
+    $provider = [System.Globalization.CultureInfo]::InvariantCulture
+    try {
+      $Status = ([System.DateTime]::ParseExact($Status, $dateFormat, $provider)).ToString($dateFormat)
```
Comment (weshaggard) [inline eng/common/scripts/ChangeLog-Operations.ps1:138]: Is there a reason that doing something like `[DateTime]$status` won't work?

File: eng/common/scripts/ChangeLog-Operations.ps1 (hunk 7 above)
Comment (weshaggard) [inline eng/common/scripts/ChangeLog-Operations.ps1:127]: Should we make $Status a datetime and let some value like Zero or $null indicate unreleased.

File: eng/common/scripts/ChangeLog-Operations.ps1 [hunk 9]
```diff
@@ -121,10 +119,117 @@ function Confirm-ChangeLogEntry {
...
+    ReleaseTitle   = $Title
+    ReleaseContent = $Content
+  }
+
+  return $newChangeLogEntry
+}
+
+function Add-ChangeLogEntry {
+  param (
+    [Parameter(Mandatory = $false)]
```
Comment (weshaggard) [inline eng/common/scripts/ChangeLog-Operations.ps1:161]: Should this be mandatory?

File: eng/common/scripts/ChangeLog-Operations.ps1 [hunk 10]
```diff
@@ -121,10 +119,117 @@ function Confirm-ChangeLogEntry {
...
+  {
+    LogError "This Changelog already contains this entry. Do you mean to edit the entry?"
+    return
+  }
+
+  $newChangeLogEntry = New-ChangeLogEntry -Version $NewEntryVersion -Status $NewEntryReleaseStatus `
+  -Content $NewEntryContent
+  $ChangeLogEntries.Add($NewEntryVersion,$newChangeLogEntry)
+
+  $ChangeLogEntries
```
Comment (weshaggard) [inline eng/common/scripts/ChangeLog-Operations.ps1:181]: Is this just for debugging?

File: eng/common/scripts/ChangeLog-Operations.ps1 [hunk 11]
```diff
@@ -121,10 +119,117 @@ function Confirm-ChangeLogEntry {
...
+    ReleaseVersion = $Version
+    ReleaseStatus  = $Status
+    ReleaseTitle   = $Title
+    ReleaseContent = $Content
+  }
+
+  return $newChangeLogEntry
+}
+
+function Add-ChangeLogEntry {
```
Comment (weshaggard) [inline eng/common/scripts/ChangeLog-Operations.ps1:159]: What value does this function really have? It seems like it would be pretty easy to add to a hash that you already have.

File: eng/common/scripts/ChangeLog-Operations.ps1 [hunk 12]
```diff
@@ -121,10 +119,117 @@ function Confirm-ChangeLogEntry {
...
+  $newChangeLogEntry = New-ChangeLogEntry -Version $NewEntryVersion -Status $NewEntryReleaseStatus `
+  -Content $NewEntryContent
+  $ChangeLogEntries.Add($NewEntryVersion,$newChangeLogEntry)
+
+  $ChangeLogEntries
+  
+  Set-ChangeLogContent -ChangeLogLocation $ChangeLogLocation -ChangeLogEntries $ChangeLogEntries
+}
+
+function Edit-ChangeLogEntry {
```
Comment (weshaggard) [inline eng/common/scripts/ChangeLog-Operations.ps1:186]: What value does this function really add? If you already have the changelog entry it seems pretty easy to set these value without needing to call this function.

File: eng/common/scripts/ChangeLog-Operations.ps1 [hunk 13]
```diff
@@ -121,10 +119,117 @@ function Confirm-ChangeLogEntry {
...
+
+function Set-ChangeLogContent {
+  param (
+    [Parameter(Mandatory = $true)]
+    [String]$ChangeLogLocation,
+    [Parameter(Mandatory = $true)]
+    $ChangeLogEntries
+  )
+
+  try {
```
Comment (weshaggard) [inline eng/common/scripts/ChangeLog-Operations.ps1:216]: Is there any harm in always just importing SemVer.ps1? I think even if you import it more than once it doesn't hurt anything.

File: eng/common/scripts/ChangeLog-Operations.ps1 [hunk 14]
```diff
@@ -1,4 +1,5 @@
 # Common Changelog Operations
+if (!(Test-Path Function:LogWarning)) { . "${PSScriptRoot}\logging.ps1" }
```
Comment (weshaggard) [inline eng/common/scripts/ChangeLog-Operations.ps1:2]: Why not just always import this and SemVer.ps1 if this script needs it? I don't thing PS will care if  this gets imported more then once.

File: eng/common/scripts/ChangeLog-Operations.ps1 [hunk 15]
```diff
@@ -121,10 +119,117 @@ function Confirm-ChangeLogEntry {
...
+  }
+  catch {
+    . "${PSScriptRoot}\SemVer.ps1"
+  }
+
+  $changeLogContent = @()
+  $changeLogContent += "# Release History"
+  $changeLogContent += ""
+
+  $VersionsSorted = [AzureEngSemanticVersion]::SortVersionStrings($ChangeLogEntries.Keys)
```
Comment (weshaggard) [inline eng/common/scripts/ChangeLog-Operations.ps1:227]: This has the potential to throw an exception if one of the versions doesn't correctly parse. We might need to handle that.

File: eng/common/scripts/Update-ChangeLog.ps1 [hunk 16]
```diff
@@ -0,0 +1,109 @@
...
+
+    $dateFormat = "yyyy-MM-dd"
+    $provider = [System.Globalization.CultureInfo]::InvariantCulture
+    if ($ReleaseDate)
```
Comment (weshaggard) [inline eng/common/scripts/Update-ChangeLog.ps1:39]: We could always make ReleaseDate be a DateTime parameter and if it set then we can just call ToString on it we shouldn't need to parse it and then call ToString.

File: eng/common/scripts/ChangeLog-Operations.ps1 [hunk 17]
```diff
@@ -121,10 +119,117 @@ function Confirm-ChangeLogEntry {
...
+    catch {
+        LogWarning "Invalid date [ $Status ] passed as status for Version [$Version]. Please use a valid date in the format '$dateFormat' or use '(Unreleased)'"
+    }
+  }
+
+  $Status = "($Status)"
+  if (!$Content) { $Content = @() }
+  if (!$Title) { $Title = "## $Version $Status" }
+
+  $newChangeLogEntry = [pscustomobject]@{
```
Comment (weshaggard) [inline eng/common/scripts/ChangeLog-Operations.ps1:149]: Should we try to parse the Version?

File: eng/common/scripts/ChangeLog-Operations.ps1 [hunk 18]
```diff
@@ -121,10 +119,117 @@ function Confirm-ChangeLogEntry {
...
+    $provider = [System.Globalization.CultureInfo]::InvariantCulture
+    try {
+      $Status = ([System.DateTime]::ParseExact($Status, $dateFormat, $provider)).ToString($dateFormat)
+    }
+    catch {
+        LogWarning "Invalid date [ $Status ] passed as status for Version [$Version]. Please use a valid date in the format '$dateFormat' or use '(Unreleased)'"
+    }
+  }
+
+  $Status = "($Status)"
```
Comment (weshaggard) [inline eng/common/scripts/ChangeLog-Operations.ps1:145]: Why add the parens `()` here? It seems like you have to remove them in various places. Maybe you should only add them when you are building the title.

File: eng/common/scripts/Update-ChangeLog.ps1 [hunk 19]
```diff
@@ -0,0 +1,109 @@
...
+                }
+            }
+        }
+
+        $PresentVersionsSorted = [AzureEngSemanticVersion]::SortVersionStrings($ChangeLogEntries.Keys)
+        $LatestVersion = $PresentVersionsSorted[0]
+
+        if ($ReplaceVersion) 
+        {
+            $NewChangeLogEntries = Edit-ChangeLogEntry -ChangeLogEntries $ChangeLogEntries -VersionToEdit $LatestVersion `
```
Comment (weshaggard) [inline eng/common/scripts/Update-ChangeLog.ps1:91]: I would just get the entry by the key and then set the values directly here.

File: eng/common/scripts/Update-ChangeLog.ps1 [hunk 20]
```diff
@@ -0,0 +1,109 @@
...
+            -NewEntryReleaseVersion $Version -NewEntryReleaseStatus $ReleaseStatus
+            Set-ChangeLogContent -ChangeLogLocation $PkgProperties.ChangeLogPath -ChangeLogEntries $NewChangeLogEntries
+        }
+        else 
+        {
+            $NewChangeLogEntries = Add-ChangeLogEntry -ChangeLogEntries $ChangeLogEntries -NewEntryVersion $Version `
```
Comment (weshaggard) [inline eng/common/scripts/Update-ChangeLog.ps1:97]: I would just create the new entry and then add it the the entries here.

File: eng/common/scripts/Update-ChangeLog.ps1 [hunk 21]
```diff
@@ -0,0 +1,109 @@
...
+            -NewEntryReleaseStatus $ReleaseStatus
+            Set-ChangeLogContent -ChangeLogLocation $PkgProperties.ChangeLogPath -ChangeLogEntries $NewChangeLogEntries
+        }
+    }
+    else 
+    {
+        $NewChangeLogEntries = Add-ChangeLogEntry -NewEntryVersion $Version -NewEntryReleaseStatus $ReleaseStatus
```
Comment (weshaggard) [inline eng/common/scripts/Update-ChangeLog.ps1:104]: I would just create the new entry here.

File: eng/common/scripts/Update-ChangeLog.ps1 [hunk 22]
```diff
@@ -0,0 +1,109 @@
...
+        Set-ChangeLogContent -ChangeLogLocation $PkgProperties.ChangeLogPath -ChangeLogEntries $NewChangeLogEntries
```
Comment (weshaggard) [inline eng/common/scripts/Update-ChangeLog.ps1:105]: Why not move the Set-ChangeLogContent function call to the end of this script instead of duplicating it in various places?

File: eng/pipelines/eng-common-sync.yml (hunk 5 above)
Comment (chidozieononiwu) [reply eng/pipelines/eng-common-sync.yml:39]: Java long paths, I will switch it back once that is fixed

File: eng/common/scripts/Update-ChangeLog.ps1 (hunk 1 above)
Comment (chidozieononiwu) [reply eng/common/scripts/Update-ChangeLog.ps1:20]: I guess it not necessary. But I was trying to make sure you can only pass the `ReleaseDate` parameter when Unreleased is set to $False. That eill make other part of the script more straight forward.

File: eng/common/scripts/Update-ChangeLog.ps1 (hunk 2 above)
Comment (chidozieononiwu) [reply eng/common/scripts/Update-ChangeLog.ps1:1]: Right

File: eng/common/scripts/Update-ChangeLog.ps1 (hunk 3 above)
Comment (chidozieononiwu) [reply eng/common/scripts/Update-ChangeLog.ps1:33]: It needed when using DynamicParameter.

File: eng/common/scripts/ChangeLog-Operations.ps1 (hunk 8 above)
Comment (chidozieononiwu) [reply eng/common/scripts/ChangeLog-Operations.ps1:138]: Assigning the string directly like this [DateTime]$Status= 2019-15-6 does not result in the correct datetime.

File: eng/common/scripts/ChangeLog-Operations.ps1 (hunk 14 above)
Comment (chidozieononiwu) [reply eng/common/scripts/ChangeLog-Operations.ps1:2]: Ok, I,ll go ahead and do that.

File: eng/common/scripts/ChangeLog-Operations.ps1 (hunk 7 above)
Comment (chidozieononiwu) [reply eng/common/scripts/ChangeLog-Operations.ps1:127]: It does not seem that that will work. Assigning the string directly like this [DateTime]$Status= 2019-15-6 does not result in the correct datetime.

File: eng/common/scripts/ChangeLog-Operations.ps1 (hunk 18 above)
Comment (chidozieononiwu) [reply eng/common/scripts/ChangeLog-Operations.ps1:145]: So far we have been adding the parenthesis.  I only removed it to verify the date. Ad that also take care of cases where it was not added.

File: eng/common/scripts/ChangeLog-Operations.ps1 (hunk 17 above)
Comment (chidozieononiwu) [reply eng/common/scripts/ChangeLog-Operations.ps1:149]: I was going to do that but I wonder if that's necessary here. We would need to determine if its a python version because there is a different parse function for that. So far the changelog logic have not been language dependent and I wanted to keep it that way.

Comment (chidozieononiwu) [reply eng/common/scripts/ChangeLog-Operations.ps1:149]: Maybe we can move SetupPythonConventions into the pythons Language setting, Then we can keep ParseVersionString in common.ps1 and only call SetupPythonConventions  if it has been loaded from python Language Settings file.

File: eng/common/scripts/ChangeLog-Operations.ps1 (hunk 12 above)
Comment (chidozieononiwu) [reply eng/common/scripts/ChangeLog-Operations.ps1:186]: Well it also calls `Set-ChangeLogContent` which provides a way to write the Hashtable of changelog Entries back to the ChangeLog file.

File: eng/common/scripts/ChangeLog-Operations.ps1 (hunk 9 above)
Comment (chidozieononiwu) [reply eng/common/scripts/ChangeLog-Operations.ps1:161]: oh, No. I want to allow starting a new changelog by not passing any previous changelog entries.

File: eng/common/scripts/copy-docs-to-blobstorage.ps1 (hunk 4 above)
Comment (weshaggard) [reply eng/common/scripts/copy-docs-to-blobstorage.ps1:234]:
```suggestion
if (Test-Path "Function:$PublishGithubIODocsFn")
```

File: eng/common/scripts/Package-Properties.ps1 [hunk 23]
```diff
@@ -91,7 +91,7 @@ function Get-PkgProperties
     {
         $pkgDirectoryPath = Join-Path $serviceDirectoryPath $directory.Name
 
-        if ((Get-ChildItem -Path Function: | ? { $_.Name -eq $GetPackageInfoFromRepoFn }).Count -gt 0)
+        if (Test-Path Function:GetPackageInfoFromRepoFn)
```
Comment (weshaggard) [inline eng/common/scripts/Package-Properties.ps1:94]:
```suggestion
        if (Test-Path "Function:$GetPackageInfoFromRepoFn")
```

File: eng/common/scripts/ChangeLog-Operations.ps1 [hunk 24]
```diff
@@ -1,4 +1,5 @@
 # Common Changelog Operations
+if (!(Test-Path Function:LogWarning)) { . "${PSScriptRoot}\logging.ps1" }
 
 $RELEASE_TITLE_REGEX = "(?<releaseNoteTitle>^\#+.*(?<version>\b\d+\.\d+\.\d+([^0-9\s][^\s:]+)?)(\s(?<releaseStatus>\(Unreleased\)|\(\d{4}-\d{2}-\d{2}\)))?)"
```
Comment (weshaggard) [inline eng/common/scripts/ChangeLog-Operations.ps1:4]:
```suggestion
$RELEASE_TITLE_REGEX = "(?<releaseNoteTitle>^\#+.*(?<version>\b\d+\.\d+\.\d+([^0-9\s][^\s:]+)?)(\s\((?<releaseStatus>Unreleased|\d{4}-\d{2}-\d{2})\))?)"
```

File: eng/common/scripts/ChangeLog-Operations.ps1 [hunk 25]
```diff
@@ -121,10 +119,117 @@ function Confirm-ChangeLogEntry {
...
+      $Status = ([System.DateTime]::ParseExact($Status, $dateFormat, $provider)).ToString($dateFormat)
+    }
+    catch {
+        LogWarning "Invalid date [ $Status ] passed as status for Version [$Version]. Please use a valid date in the format '$dateFormat' or use '(Unreleased)'"
+    }
+  }
+
+  $Status = "($Status)"
+  if (!$Content) { $Content = @() }
+  if (!$Title) { $Title = "## $Version $Status" }
```
Comment (weshaggard) [inline eng/common/scripts/ChangeLog-Operations.ps1:147]:
```suggestion
  if (!$Title) { $Title = "## $Version ($Status)" }
```

File: eng/common/scripts/Package-Properties.ps1 (hunk 23 above)
Comment (chidozieononiwu) [reply eng/common/scripts/Package-Properties.ps1:94]: This seems to always return true. This won't work because its actually a variable. I'll leave the previous logic.

File: eng/common/scripts/ChangeLog-Operations.ps1 (hunk 18 above)
Comment (chidozieononiwu) [reply eng/common/scripts/ChangeLog-Operations.ps1:145]: I think its better to continue with this, or else I will have to update a lot of scripts where we are already consuming the $ReleaseStatus with the parenthesis.

File: eng/common/scripts/copy-docs-to-blobstorage.ps1 (hunk 4 above)
Comment (chidozieononiwu) [reply eng/common/scripts/copy-docs-to-blobstorage.ps1:234]: This actually does not work. It always returns $True

File: eng/common/scripts/ChangeLog-Operations.ps1 (hunk 18 above)
Comment (weshaggard) [reply eng/common/scripts/ChangeLog-Operations.ps1:147]: Are there scripts outside of the scope of this PR that would need to be changed? If not then I still think that it might be better to make the change and assume there is no parenthesis around the status as you are removing and adding in various places which makes it complex to understand when they should be or should not be there.

File: eng/common/scripts/ChangeLog-Operations.ps1 [hunk 26]
```diff
@@ -121,10 +122,88 @@ function Confirm-ChangeLogEntry {
...
+    [String]$Version,
+    [String]$Status=$UNRELEASED_TAG,
+    [String]$Title,
+    [String[]]$Content,
+    [String]$IgnoreInvalids=$True
+  )
+
+  # Validate RelaseStatus
+  $Status = $Status.Trim().Trim("()")
+  if ($Status -ne "Unreleased") {
```
Comment (weshaggard) [inline eng/common/scripts/ChangeLog-Operations.ps1:138]: This is another place we could share the constant if we didn't assume there were parenthesis around the status.

File: eng/common/scripts/SemVer.ps1 [hunk 27]
```diff
@@ -22,6 +22,7 @@ class AzureEngSemanticVersion {
   [string] $RawVersion
   [bool] $IsSemVerFormat
   [string] $DefaultPrereleaseLabel
+  [string] $ParseLanguage = $Language
```
Comment (weshaggard) [inline eng/common/scripts/SemVer.ps1:25]: This should probably be a static variable so that we can control outside the scope of an individual object.

File: eng/common/scripts/SemVer.ps1 [hunk 28]
```diff
@@ -143,7 +144,14 @@ class AzureEngSemanticVersion {
 
   static [string[]] SortVersionStrings([string[]] $versionStrings)
   {
-    $versions = $versionStrings | ForEach-Object { [AzureEngSemanticVersion]::ParseVersionString($_) }
```
Comment (weshaggard) [inline eng/common/scripts/SemVer.ps1:146]: We shouldn't do this update at this level but instead in the constructor we should set the convention to python if the ParseLanguage is python.

File: eng/common/scripts/ChangeLog-Operations.ps1 [hunk 29]
```diff
@@ -121,10 +122,88 @@ function Confirm-ChangeLogEntry {
...
+        LogWarning "Invalid date [ $Status ] passed as status for Version [$Version]. Please use a valid date in the format '$DATE_FORMAT' or use '$UNRELEASED_TAG'"
+        if (!$IgnoreInvalids) { exit 1 }
+    }
+  }
+  $Status = "($Status)"
+
+  # Validate Version
+  try {
+    if ($Language -eq "python") {
+      $Version = ([AzureEngSemanticVersion]::ParsePythonVersionString($Version)).ToString()
```
Comment (weshaggard) [inline eng/common/scripts/ChangeLog-Operations.ps1:152]: Once you do the python detection in the constructor you shouldn't need to have this fork here.

File: eng/common/scripts/ChangeLog-Operations.ps1 [hunk 30]
```diff
@@ -121,10 +122,88 @@ function Confirm-ChangeLogEntry {
...
+
+  # Validate RelaseStatus
+  $Status = $Status.Trim().Trim("()")
+  if ($Status -ne "Unreleased") {
+    try {
+      $Status = ([DateTime]$Status).ToString($DATE_FORMAT)
+    }
+    catch {
+        LogWarning "Invalid date [ $Status ] passed as status for Version [$Version]. Please use a valid date in the format '$DATE_FORMAT' or use '$UNRELEASED_TAG'"
+        if (!$IgnoreInvalids) { exit 1 }
```
Comment (weshaggard) [inline eng/common/scripts/ChangeLog-Operations.ps1:144]: This pattern of ignoreInvalids is a little odd. Instead I would suggest we simply return $null and let the caller decide what to do.

File: eng/common/scripts/Update-ChangeLog.ps1 [hunk 31]
```diff
@@ -0,0 +1,98 @@
...
+  [String]$ReleaseDate
+)
+
+if ($ReleaseDate -and ($Unreleased -eq $True)) {
+    LogError "Do not pass 'ReleaseDate' arguement when 'Unreleased' is true"
+    exit 1
+}
+
+. "${PSScriptRoot}\common.ps1"
+$UNRELEASED_TAG = "(Unreleased)"
```
Comment (weshaggard) [inline eng/common/scripts/Update-ChangeLog.ps1:27]: We shouldn't need to redefine these here.

File: eng/common/scripts/ChangeLog-Operations.ps1 [hunk 32]
```diff
@@ -1,6 +1,10 @@
 # Common Changelog Operations
+. "${PSScriptRoot}\logging.ps1"
+. "${PSScriptRoot}\SemVer.ps1"
 
 $RELEASE_TITLE_REGEX = "(?<releaseNoteTitle>^\#+.*(?<version>\b\d+\.\d+\.\d+([^0-9\s][^\s:]+)?)(\s(?<releaseStatus>\(Unreleased\)|\(\d{4}-\d{2}-\d{2}\)))?)"
+$UNRELEASED_TAG = "(Unreleased)"
+$DATE_FORMAT = "yyyy-MM-dd"
```
Comment (weshaggard) [inline eng/common/scripts/ChangeLog-Operations.ps1:7]: Given this will be at the global variable scoped we should update the names to be a little more specific. Maybe something like "CHANGELOG_DATE_FORMAT" and "CHANGELOG_UNRELEASED_STATUS"

File: eng/common/scripts/copy-docs-to-blobstorage.ps1 (hunk 4 above)
Comment (weshaggard) [reply eng/common/scripts/copy-docs-to-blobstorage.ps1:234]: I believe your current code will always return false. From my local testing I think you need to add the quotes and have the "$PublishGithubIODocsFn" expand the name of the function we are checking.

File: eng/common/scripts/Invoke-GitHubAPI.ps1 [hunk 33]
```diff
@@ -1,4 +1,4 @@
-if ((Get-ChildItem -Path Function: | ? { $_.Name -eq "LogWarning" }).Count -eq 0) {
+if (!(Test-Path Function:LogWarning)) {
```
Comment (weshaggard) [inline eng/common/scripts/Invoke-GitHubAPI.ps1:1]: Might as well remove this and always import logging.ps1 like you have done in other places.

File: eng/common/scripts/Update-ChangeLog.ps1 [hunk 34]
```diff
@@ -0,0 +1,98 @@
...
+
+        if ($Unreleased -and ($ChangeLogEntries[$Version].ReleaseStatus -ne $ReleaseStatus))
+        {
+            LogWarning "Version is already present in change log with a release date. Please review [$($PkgProperties.ChangeLogPath)]"
+            exit(0)
+        }
+
+        if (!$Unreleased -and ($ChangeLogEntries[$Version].ReleaseStatus -ne $UNRELEASED_TAG))
+        {
+            if (Get-Date ($ChangeLogEntries[$Version].ReleaseStatus).Trim("()") -gt Get-Date $ReleaseStatus.Trim("()"))
```
Comment (weshaggard) [inline eng/common/scripts/Update-ChangeLog.ps1:70]: Is there any differences between calling "Get-Date" vs casting to DateTime like you are doing in the changelog-operations.ps1 file? We should try and be consistent so things work anywhere we are parsing or formatting the date. Perhaps that is worth a helper function.

File: eng/common/scripts/copy-docs-to-blobstorage.ps1 (hunk 4 above)
Comment (chidozieononiwu) [reply eng/common/scripts/copy-docs-to-blobstorage.ps1:234]: I thought I returned to the original logic. Even with the quotes this does not work because `$PublishGithubIODocsFn` is actually a variable.

File: eng/common/scripts/ChangeLog-Operations.ps1 (hunk 18 above)
Comment (chidozieononiwu) [reply eng/common/scripts/ChangeLog-Operations.ps1:147]: Yes there are a couple of scripts that are already using this. Specifically the new collect changelog scripts in the language repos.

File: eng/common/scripts/copy-docs-to-blobstorage.ps1 (hunk 4 above)
Comment (weshaggard) [reply eng/common/scripts/copy-docs-to-blobstorage.ps1:234]: Can you please test it again as I was able to get it to work with quotes because the variable would result in a full string that would be the function location that Test-Path would be able to understand.

File: eng/common/scripts/ChangeLog-Operations.ps1 (hunk 18 above)
Comment (weshaggard) [reply eng/common/scripts/ChangeLog-Operations.ps1:147]: OK that make this a leaky abstraction because now all consumers have to know about and handle these parenthesis. Perhaps we can consider improvements when we work on those scripts.

File: eng/common/scripts/copy-docs-to-blobstorage.ps1 (hunk 4 above)
Comment (chidozieononiwu) [reply eng/common/scripts/copy-docs-to-blobstorage.ps1]: It always returns true. Even when the variable is not in scope.

File: eng/common/scripts/ChangeLog-Operations.ps1 (hunk 30 above)
Comment (chidozieononiwu) [reply eng/common/scripts/ChangeLog-Operations.ps1:144]: Well, this is going to be a problem. Right now not all entries in our changelogs have correct version. I was trying to have a way to detect bad version yet no have to break the process when reading ChangeLog entries.

File: eng/common/scripts/Update-ChangeLog.ps1 (hunk 34 above)
Comment (chidozieononiwu) [reply eng/common/scripts/Update-ChangeLog.ps1:68]: All that needed here is to make sure the new date is not behind, I don't think it matters in what format the date is in.

File: eng/common/scripts/Update-ChangeLog.ps1 [hunk 35]
```diff
@@ -0,0 +1,119 @@
...
+    exit(0)
+}
+
+$PkgProperties = Get-PkgProperties -PackageName $PackageName -ServiceDirectory $ServiceDirectory
+$ChangeLogEntries = Get-ChangeLogEntries -ChangeLogLocation $PkgProperties.ChangeLogPath
+
+
+if ($ChangeLogEntries.Contains($Version))
+{
+    if ($ChangeLogEntries[$Version].ReleaseStatus -eq $ReleaseStatus)
```
Comment (weshaggard) [inline eng/common/scripts/Update-ChangeLog.ps1:58]: At some point we should set the indention level to 2 spaces like the other ps files.
